### PR TITLE
[IMP] crm claim

### PR DESCRIPTION
**Pull request by hmo-odoo**
Task: https://www.odoo.com/web#id=9372&view_type=form&model=project.task&menu_id=3942&action=327

Pad: https://pad.odoo.com/p/r.jhh4fJAi0f6GvlJ8

### CODE REVIEW

File: addons/crm_claim/__openerp__.py [hunk 1]
```diff
@@ -19,12 +19,14 @@
...
     'data': [
+        'crm_claim_data.xml',
         'crm_claim_view.xml',
         'crm_claim_menu.xml',
         'security/ir.model.access.csv',
         'report/crm_claim_report_view.xml',
-        'crm_claim_data.xml',
         'res_partner_view.xml',
+        'crm_claim_tip.xml',
+        'crm_claim_action_rule_demo.xml'
```
Comment (atp-odoo) [inline addons/crm_claim/__openerp__.py]: declare this file into demo tag.

File: addons/crm_claim/crm_claim.py [hunk 2]
```diff
@@ -38,13 +43,63 @@ class crm_claim(osv.osv):
...
+        # restore order of the search
+        result.sort(lambda x,y: cmp(stage_ids.index(x[0]), stage_ids.index(y[0])))
+
+        fold = {}
+        for stage in Stage.browse(cr, access_rights_uid, stage_ids, context=context):
+            fold[stage.id] = stage.fold or False
+        return result, fold
+
+    def action_calendar_followup(self, cr, uid, ids, context=None):
+        action = self.pool.get('ir.actions.act_window').for_xml_id(cr, uid, 'calendar', 'action_calendar_event', context)
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim.py]: use self.pool['ir.actions.act_window'] instead of self.pool.get('ir.actions.act_window')

File: addons/crm_claim/calendar_event.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+from openerp import api, fields, models
+
+class CalendarEvent(models.Model):
+    _inherit = "calendar.event"
+
+    claim_id = fields.Many2one('crm.claim', string="Claim", ondelete='cascade')
+    start_datetime = fields.Datetime(default= fields.Datetime.now())
+
+    @api.onchange('start_datetime', 'stop_datetime')
```
Comment (atp-odoo) [inline addons/crm_claim/calendar_event.py]: remove this onchange not required.

File: addons/crm_claim/crm_claim.py [hunk 4]
```diff
@@ -72,25 +127,59 @@ def _get_default_stage_id(self, cr, uid, context=None):
...
     }
 
+    _group_by_full = {
+        'stage_id': _read_group_stage_ids
+    }
+
+    def _track_subtype(self, cr, uid, ids, init_values, context=None):
+        record = self.browse(cr, uid, ids[0], context=context)
+        if 'stage_id' in init_values and record.stage_id.name == 'Settled':
+            return 'crm_claim.mt_claim_settled_'
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim.py]: mt_claim_settled_ ? id should be mt_claim_settled

File: addons/crm_claim/crm_claim.py [hunk 5]
```diff
@@ -72,25 +127,59 @@ def _get_default_stage_id(self, cr, uid, context=None):
...
+        elif 'stage_id' in init_values and record.stage_id.name == 'Rejected':
+            return 'crm_claim.mt_claim_rejected_'
+        return super(crm_claim, self)._track_subtype(cr, uid, ids, init_values, context=context)
+
+    def action_get_attachment_view(self, cr, uid, ids, context=None):
+        IrModelData = self.pool['ir.model.data']
+        # Select the view
+        kanban_view_id = IrModelData.xmlid_to_res_id(cr, uid, 'mail.view_document_file_kanban')
+        form_view_id = IrModelData.xmlid_to_res_id(cr, uid, 'base.view_attachment_form')
+        tree_view_id = IrModelData.xmlid_to_res_id(cr, uid, 'base.view_attachment_tree')
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim.py]: Not used . remote it

File: addons/crm_claim/crm_claim.py [hunk 6]
```diff
@@ -171,7 +288,7 @@ class res_partner(osv.osv):
     def _claim_count(self, cr, uid, ids, field_name, arg, context=None):
         Claim = self.pool['crm.claim']
         return {
-            partner_id: Claim.search_count(cr,uid, [('partner_id', '=', partner_id)], context=context)  
+            partner_id: Claim.search_count(cr, uid, [('partner_id', '=', partner_id)], context=context)
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim.py]: unused diff. remove this change.

File: addons/crm_claim/crm_claim_menu.xml [hunk 7]
```diff
@@ -47,6 +53,7 @@
         </record>
 
 
+
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim_menu.xml]: remove unused change to remove diff.

File: addons/crm_claim/crm_claim_view.xml [hunk 8]
```diff
@@ -1,229 +1,291 @@
...
-                        <field name="sequence"/>
+                    <field name="case_default"/>
+                    <field name="sequence"/>
+                     <field name="on_followup"/>
+                </group>
+            </form>
+        </field>
+    </record>
+
+    <record id="crm_claim_stage_act" model="ir.actions.act_window">
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim_view.xml]: change crm_claim_stage_action based on new guideline.

File: addons/crm_claim/crm_claim_view.xml [hunk 9]
```diff
@@ -1,229 +1,291 @@
...
+            claim entered in the system. The stages define all the steps
+            required for the resolution of a claim.
+          </p>
+        </field>
+    </record>
+
+    <!-- Claims    -->
+
+    <record model="ir.ui.view" id="crm_case_claims_tree_view">
+        <field name="name">CRM - Claims Tree</field>
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim_view.xml]: change name : crm.claim.tree

File: addons/event/event_data.xml [hunk 10]
```diff
@@ -13,6 +13,12 @@
             <field name="state">open</field>
         </record>
 
+        <!-- Requests Links -->
+        <record id="req_link_event" model="res.request.link">
```
Comment (atp-odoo) [inline addons/event/event_data.xml]: why this change? unused change. remove it

File: addons/crm_claim/__openerp__.py [hunk 11]
```diff
@@ -19,14 +19,18 @@
...
         'crm_claim_menu.xml',
         'security/ir.model.access.csv',
-        'report/crm_claim_report_view.xml',
-        'crm_claim_data.xml',
+        'report/crm_claim_report_view.xml'
+    ],
+    'demo': [
+        'crm_claim_demo.xml',
         'res_partner_view.xml',
+        'crm_claim_tip.xml',
```
Comment (atp-odoo) [inline addons/crm_claim/__openerp__.py]: crm_claim_tip should be in data tag.

File: addons/crm_claim/calendar_event.py [hunk 12]
```diff
@@ -0,0 +1,10 @@
+# -*- coding: utf-8 -*-
+
+import datetime
+
+from openerp import api, fields, models
```
Comment (atp-odoo) [inline addons/crm_claim/calendar_event.py]: unused api import.

File: addons/crm_claim/crm_claim.py [hunk 13]
```diff
@@ -123,17 +210,44 @@ def onchange_partner_id(self, cr, uid, ids, partner_id, email=False, context=Non
...
-            return {'value': {'email_from': False, 'partner_phone': False}}
+            return {'value': {'email_from': False, 'partner_phone': False, 'mobile': False}}
         address = self.pool.get('res.partner').browse(cr, uid, partner_id, context=context)
-        return {'value': {'email_from': address.email, 'partner_phone': address.phone}}
+        return {'value': {'email_from': address.email, 'partner_phone': address.phone, 'mobile': address.mobile}}
+
+    def create_claim_followup(self, cr, uid, ids, stage_id, context=None):
+        CrmClaimFollowup = self.pool['crm.claim.followup']
+        stage = self.pool['crm.claim.stage'].browse(cr, uid, stage_id)
+        if not stage.on_followup:
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim.py]: put comment what's purpose of on_followup field.

File: addons/crm_claim/crm_claim.py [hunk 14]
```diff
@@ -171,7 +285,7 @@ class res_partner(osv.osv):
     def _claim_count(self, cr, uid, ids, field_name, arg, context=None):
         Claim = self.pool['crm.claim']
         return {
-            partner_id: Claim.search_count(cr,uid, [('partner_id', '=', partner_id)], context=context)  
+            partner_id: Claim.search_count(cr,uid, [('partner_id', '=', partner_id)], context=context)
```
Comment (atp-odoo) [inline addons/crm_claim/crm_claim.py]: remove unused diff

File: addons/event/event_data.xml [hunk 15]
```diff
@@ -12,7 +12,7 @@
             <field name="action_id" ref="action_client_event_menu"/>
             <field name="state">open</field>
         </record>
-
+
```
Comment (atp-odoo) [inline addons/event/event_data.xml]: unused diff

File: addons/crm_claim/crm_claim_view.xml [hunk 16]
```diff
@@ -1,229 +1,291 @@
...
+    <!-- Crm claim Search view -->
+
+    <record id="view_crm_case_claims_filter" model="ir.ui.view">
+        <field name="name">CRM - Claims Search</field>
+        <field name="model">crm.claim</field>
+        <field name="arch" type="xml">
+            <search string="Search Claims">
+                <field name="name" string="Claims"/>
+                <separator/>
+                <filter string="Unassigned Claims" icon="terp-personal-" domain="[('user_id','=', False)]" help="Unassigned Claims" />
```
Comment (hmo-odoo) [inline addons/crm_claim/crm_claim_view.xml]: remove icon from search view. it's useless

File: addons/crm_claim/crm_claim_view.xml [hunk 17]
```diff
@@ -1,229 +1,291 @@
...
+                <field name="user_id"/>
+                <field name="stage_id"/>
+                <group expand="0" string="Group By">
+                    <filter string="Partner" icon="terp-partner" domain="[]" help="Partner" context="{'group_by':'partner_id'}"/>
+                    <filter string="Responsible" icon="terp-personal" domain="[]" help="Responsible User" context="{'group_by':'user_id'}"/>
+                    <filter string="Stage" icon="terp-stage" domain="[]" context="{'group_by':'stage_id'}"/>
+                    <filter string="Type" icon="terp-stock_symbol-selection" domain="[]" context="{'group_by':'categ_id'}"/>
+                    <filter string="Claim Month" icon="terp-go-month" domain="[]" help="Claim Date by Month" context="{'group_by':'date'}"/>
+                    <filter string="Deadline" icon="terp-go-month" domain="[]" context="{'group_by':'date_deadline'}"/>
+                    <filter string="Closure" icon="terp-go-month" domain="[]" help="Date Closed" context="{'group_by':'date_closed'}"  groups="base.group_no_one" invisible="1"/>
```
Comment (hmo-odoo) [inline addons/crm_claim/crm_claim_view.xml]: why invisible=1 ? if not use, can you remove filter ?

File: addons/crm_claim/crm_claim_view.xml [hunk 18]
```diff
@@ -1,229 +1,291 @@
...
+                                <field name="email_from" widget="email"/>
+                                <field name="mobile"/>
+                            </group>
+                            <group string="Claim Subject" groups="base.group_user">
+                                <field name="ref"/>
+                            </group>
+                            <field name="description" colspan="4" nolabel="1" groups="base.group_user" placeholder="Description..."/>
+                        </page>
+                        <page string="Action" groups="base.group_user">
+                           <separator colspan="4" string="Follow-up"/>
```
Comment (hmo-odoo) [inline addons/crm_claim/crm_claim_view.xml]: indentation issue here

File: addons/crm_claim/crm_claim_followup.py [hunk 19]
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
+
+from openerp import api, fields, models
```
Comment (hmo-odoo) [inline addons/crm_claim/crm_claim_followup.py]: why import api ? it is not use. can you remove it ?

File: addons/crm_claim/crm_claim.py [hunk 20]
```diff
@@ -123,17 +210,45 @@ def onchange_partner_id(self, cr, uid, ids, partner_id, email=False, context=Non
...
         if not partner_id:
-            return {'value': {'email_from': False, 'partner_phone': False}}
+            return {'value': {'email_from': False, 'partner_phone': False, 'mobile': False}}
         address = self.pool.get('res.partner').browse(cr, uid, partner_id, context=context)
-        return {'value': {'email_from': address.email, 'partner_phone': address.phone}}
+        return {'value': {'email_from': address.email, 'partner_phone': address.phone, 'mobile': address.mobile}}
+
+    def create_claim_followup(self, cr, uid, ids, stage_id, context=None):
+        CrmClaimFollowup = self.pool['crm.claim.followup']
+        stage = self.pool['crm.claim.stage'].browse(cr, uid, stage_id)
```
Comment (hmo-odoo) [inline addons/crm_claim/crm_claim.py]: pass context in browse method.

File: addons/crm_claim/crm_claim.py [hunk 21]
```diff
@@ -123,17 +210,45 @@ def onchange_partner_id(self, cr, uid, ids, partner_id, email=False, context=Non
...
+        # on_followup: To find a particular stage that is new,settled,rejected
+        if not stage.on_followup:
+            return False
+        for claim in self.browse(cr, uid, ids, context=context):
+            CrmClaimFollowup.create(cr, uid, {'date': fields.date.today(),
```
Comment (hmo-odoo) [inline addons/crm_claim/crm_claim.py]: arrange value dict in proper format.

File: addons/crm_claim/crm_claim.py [hunk 22]
```diff
@@ -38,13 +43,63 @@ class crm_claim(osv.osv):
     _name = "crm.claim"
     _description = "Claim"
     _order = "priority,date desc"
-    _inherit = ['mail.thread']
+    _inherit = ['mail.thread', 'ir.needaction_mixin']
 
     def _get_default_stage_id(self, cr, uid, context=None):
-        """ Gives default stage_id """
+        if context is None:
```
Comment (hmo-odoo) [inline addons/crm_claim/crm_claim.py]: here you are not doing any operation in context. so I think, you can remove this stuff

File: addons/crm_claim/crm_claim_followup.py (hunk 19 above)
Comment (mpa-odoo) [inline addons/crm_claim/crm_claim_followup.py]: okay, i remove it.

**Comment by hmo-odoo** [comment]
crm_claim module is not present in master
